File: 0994-rotting-oranges-腐烂的橘子/0994.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        goodOrgs = {}
        badOrgs = {}

        h = len(grid)
        if h == 0: return
        w = len(grid[0])
        if w == 0: return

        for y in range(h):
            for x in range(w):
                if grid[y][x] == 1:
                    goodOrgs[(y,x)] = None
                elif grid[y][x] == 2:
                    badOrgs[(y,x)] = None

        # 没有好橘子
        if len(goodOrgs) == 0:
                return 0

        time = 0
        lastGoodOrgsNumber = -1
        while len(goodOrgs) > 0:
            # print("len:", len(goodOrgs), lastGoodOrgsNumber)
            if lastGoodOrgsNumber == len(goodOrgs):
                # 无法腐烂完所有橘子
                return -1

            # 初始化本次腐烂过程
            lastGoodOrgsNumber = len(goodOrgs)
            newBadOrgs = {}

            for y, x in badOrgs.keys():
                # print(y, x)
                delGoodOrgs = []
                if y > 0:
                    if (y-1, x) in goodOrgs.keys():
                        delGoodOrgs.append((y-1, x))
                        newBadOrgs[(y-1,x)] = None
                if y < h - 1:
                    if (y+1, x) in goodOrgs.keys():
                        delGoodOrgs.append((y+1, x))
                        newBadOrgs[(y+1,x)] = None
                if x > 0:
                    if (y, x-1) in goodOrgs.keys():
                        delGoodOrgs.append((y, x-1))
                        newBadOrgs[(y,x-1)] = None
                if x < w - 1:
                    if (y, x+1) in goodOrgs.keys():
                        delGoodOrgs.append((y, x+1))
                        newBadOrgs[(y,x+1)] = None
                for (y, x) in delGoodOrgs:
                    goodOrgs.pop((y, x))
            badOrgs = newBadOrgs
            time += 1
                
        return time
                

File: 0994-rotting-oranges-腐烂的橘子/0994.cpp
using std::map;
using std::pair;
using std::make_pair;

class Solution {
    
public:
    int orangesRotting(vector<vector<int>>& grid) {
        map<pair<int, int>, bool*> goodOrgs;
        map<pair<int, int>, bool*> badOrgs;
        
        int h = grid.size();
        if (h == 0) return 0;
        int w = grid[0].size();
        if (w == 0) return 0;

        for(int y = 0; y != h; ++y) {
            for(int x = 0; x != w; ++x) {
                if (grid[y][x] == 1) {
                    goodOrgs[make_pair(y, x)] = NULL;
                }
                else if (grid[y][x] == 2) {
                    badOrgs[make_pair(y, x)] = NULL;
                }
            }
        }

        int time = 0;
        int lastGoodOrgsNumber = -1;
        while(goodOrgs.size() > 0) {
            if (lastGoodOrgsNumber == goodOrgs.size()) return -1;

            lastGoodOrgsNumber = goodOrgs.size();
            map<pair<int, int>, bool*> newBadOrgs;

            for (auto it = badOrgs.begin(); it != badOrgs.end(); ++it) {
                auto y = (*it).first.first;
                auto x = (*it).first.second;
                std::vector<pair<int, int>> delGoodOrgs;
                if (y > 0) {
                    if (goodOrgs.find(make_pair(y-1, x)) != goodOrgs.end()) {
                        delGoodOrgs.push_back(make_pair(y-1, x));
                        newBadOrgs[make_pair(y-1, x)] = NULL;
                    }
                }
                if (y < h-1) {
                    if (goodOrgs.find(make_pair(y+1, x)) != goodOrgs.end()) {
                        delGoodOrgs.push_back(make_pair(y+1, x));
                        newBadOrgs[make_pair(y+1, x)] = NULL;
                    }
                }
                if (x > 0) {
                    if (goodOrgs.find(make_pair(y, x-1)) != goodOrgs.end()) {
                        delGoodOrgs.push_back(make_pair(y, x-1));
                        newBadOrgs[make_pair(y, x-1)] = NULL;
                    }
                }
                if (x < w-1) {
                    if (goodOrgs.find(make_pair(y, x+1)) != goodOrgs.end()) {
                        delGoodOrgs.push_back(make_pair(y, x+1));
                        newBadOrgs[make_pair(y, x+1)] = NULL;
                    }
                }
                for (const auto &a: delGoodOrgs) {
                    goodOrgs.erase(a);
                }
            }
            badOrgs = newBadOrgs;
            time++;
        }

        return time;
         
    }
};